cli/src/config.rs

```diff
@@ -216,6 +218,7 @@ pub struct LayeredConfigs {
     repo: Option<config::Config>,
     env_overrides: config::Config,
     arg_overrides: Option<config::Config>,
+    builtin_overrides: config::Config,
```

Comment:
As you pointed out earlier, it doesn't break anything if we don't have it.

The two possible outcomes are:
  - No builtin layer: User sees warning, `jj` behaviour is inconsistent. 
  - Builtin layer: User sees warning, but `jj` behaves consistently.

I do think consistent behaviour is a good thing to strive for as it removes a class bugs and simplifies the user experience.

All that said, I'm starting to wonder if the better thing here is to remove the possibility of shadowing/redefining arbitrary builtins completely?

The main problem is that all these layers populate a single namespace.

Supposing we reserved the anonymous namespace for builtins that can't be shadowed, and introduced a 'user' namespace that the user is allowed to shadow/redefine.

We would provide default definitions (same as today) for things like 'user.trunk()' and 'user.immutable_heads()', but 'mutable()' and 'immutable()' would be in the anonymous namespace and not overwritable.

The thing that would make this all work though is that name resolution doesn't need the namespace and we search the anonymous namespace before we search the user namespace. Also, users can still disambiguate if they need to by specifying the fully qualified name.

What I'm doing here with the builtin layer is a hack, which I believe is why it feels wrong to you. Stepping back and introducing namespaces could be a better approach.

wdyt?

I could always submit this, and while we figure out a better approach or I could drop this if a better approach is obvious.

Reply:
> The two possible outcomes are:
> 
> * No builtin layer: User sees warning, `jj` behaviour is inconsistent.
> * Builtin layer: User sees warning, but `jj` behaves consistently.

To me, the latter is less consistent because defaults should be overridden by user, and builtins are more or less defaults.

> All that said, I'm starting to wonder if the better thing here is to remove the possibility of shadowing/redefining arbitrary builtins completely?

The major problem of disabling shadowing is that revset symbols include tags and branches, so the "builtin" symbols set can be updated dynamically by `jj git fetch`. It's weird if my "foo" alias got overridden by "foo" tag crated by someone else.

Function aliases doesn't have this problem, but they can still get broken if we add new builtin functions. Warning seems okay in that situation, but silent behavior change is bad.

Reply:
Ok. I think I have a much better understanding now. Thanks for the explanations.

I've done the following:

- Defined `builtin_immutable_heads()` in the `default` layer. which means it can indeed be redefined.
- Added warnings when the user redefines `builtin_immutable_heads()`, `immutable()` or `mutable()`. The warning points them at `immutable_heads()` as the recommended function to redefine.